airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/implementor/ProcessRecordsTask.kt

```diff
@@ -36,81 +43,104 @@ interface ProcessRecordsTask : ImplementorScope
-        val log = KotlinLogging.logger {}
-
-        log.info { "Fetching stream loader for $streamDescriptor" }
-        val streamLoader = syncManager.getOrAwaitStreamLoader(streamDescriptor)
+        registration.use {
+            inputQueue.consume().collect { (streamDescriptor, file) ->
+                log.info { "Fetching stream loader for $streamDescriptor" }
+                val streamLoader = syncManager.getOrAwaitStreamLoader(streamDescriptor)
+                val acc =
+                    accumulators.getOrPut(streamDescriptor) {
+                        streamLoader.createBatchAccumulator()
+                    }
+                log.info { "Processing records from $file for stream $streamDescriptor" }
+                val batch =
+                    try {
+                        file.localFile.inputStream().use { inputStream ->
+                            val records = inputStream.toRecordIterator()
+                            val batch = acc.processRecords(records, file.totalSizeBytes)
+                            log.info { "Finished processing $file" }
+                            batch
+                        }
+                    } finally {
+                        log.info { "Processing completed, deleting $file" }
+                        file.localFile.toFile().delete()
+                        diskManager.release(file.totalSizeBytes)
+                    }
 
-        log.info { "Processing records from $file" }
-        val batch =
-            try {
-                file.localFile.inputStream().use { inputStream ->
-                    val records =
-                        inputStream
-                            .lineSequence()
-                            .map {
-                                when (val message = deserializer.deserialize(it)) {
-                                    is DestinationStreamAffinedMessage -> message
-                                    else ->
-                                        throw IllegalStateException(
-                                            "Expected record message, got ${message::class}"
-                                        )
-                                }
-                            }
-                            .takeWhile {
-                                it !is DestinationRecordStreamComplete &&
-                                    it !is DestinationRecordStreamIncomplete
-                            }
-                            .map { it as DestinationRecord }
-                            .iterator()
-                    streamLoader.processRecords(records, file.totalSizeBytes)
+                val wrapped = BatchEnvelope(batch, file.indexRange, streamDescriptor)
+                log.info { "Updating batch $wrapped for $streamDescriptor" }
+                taskLauncher.handleNewBatch(streamDescriptor, wrapped)
+                if (batch.state != Batch.State.COMPLETE && batch.groupId == null) {
+                    outputQueue.publish(wrapped)
```

Comment:
This isn't looping. This is process records task -> process batch task. It would only loop if process batch task pushed back onto the queue. But I explicitly avoided that.

Reply:
Ah, this if check is handling the case when the processRecords is directly uploading.